src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -81,15 +81,35 @@ cbTrackingObjectFree(jvmtiEnv* jvmti_env, jlong tag)
 struct bag *
 classTrack_processUnloads(JNIEnv *env)
 {
-    debugMonitorEnter(classTrackLock);
-    if (deletedSignatures == NULL) {
-        // Class tracking not initialized, nobody's interested.
-        debugMonitorExit(classTrackLock);
-        return NULL;
-    }
-    struct bag* deleted = deletedSignatures;
-    deletedSignatures = bagCreateBag(sizeof(char*), 10);
-    debugMonitorExit(classTrackLock);
+    struct bag* new_bag = NULL;
+    struct bag* deleted = NULL;
+    jboolean retry = JNI_FALSE;
+    do {
+      // Avoid unnecessary allocations when class track has yet been activated.
+      if (deletedSignatures != NULL) {
```

Comment:
It's now kind of hard to read what used to be the `deletedSignatures == NULL` case. I think you can first just check for NULL outside the lock and then return. It will make it clear that it is very low overhead in this case.